toponetx/algorithms/exterior_calculus/metric.py

```diff
@@ -0,0 +1,756 @@
+
+
+def _grad_barycentric_3d(
+    p0: np.ndarray, p1: np.ndarray, p2: np.ndarray
+) -> tuple[np.ndarray, np.ndarray, np.ndarray, float]:
+    """Compute gradients of P1 basis functions on an R^3 triangle.
+
+    Parameters
+    ----------
+    p0 : np.ndarray
+        First vertex position, shape (3,).
+    p1 : np.ndarray
+        Second vertex position, shape (3,).
+    p2 : np.ndarray
+        Third vertex position, shape (3,).
+
+    Returns
+    -------
+    tuple[np.ndarray, np.ndarray, np.ndarray, float]
+        Gradients (g0, g1, g2) each shape (3,), and det_norm = ||(p1-p0)x(p2-p0)||.
+        The triangle area is A = 0.5 * det_norm.
+    """
+    e01 = p1 - p0
+    e02 = p2 - p0
+    n = np.cross(e01, e02)
+    nn = float(n @ n)
+    det_norm = float(np.linalg.norm(n))
+    if nn < 1e-28:
+        z = np.zeros(3, dtype=float)
+        return z, z, z, det_norm
+
+    g0 = np.cross(n, p2 - p1) / nn
+    g1 = np.cross(n, p0 - p2) / nn
+    g2 = np.cross(n, p1 - p0) / nn
+    return g0.astype(float), g1.astype(float), g2.astype(float), det_norm
+
+
+@dataclass(frozen=True)
+class MetricSpec:
+    """Describe how to build diagonal Hodge stars and anisotropic FEM tensors.
+
+    Attributes
+    ----------
+    preset : {"identity","diagonal","barycentric_lumped","circumcentric","voronoi","euclidean"}
+        Preset name controlling the backend.
+    diagonal_weights : dict[int, np.ndarray] | None
+        Explicit diagonal entries for the Hodge star at each degree k.
+        Each entry must have length equal to the number of k-simplices.
+    primal_measures : dict[int, np.ndarray] | None
+        Primal measures for generic diagonal DEC stars.
+    dual_measures : dict[int, np.ndarray] | None
+        Dual measures for generic diagonal DEC stars.
+    tensors : np.ndarray | None
+        Per-triangle SPD tensors for anisotropic stiffness, shape (nT, 3, 3).
+    fn : callable | None
+        Per-triangle tensor function fn(t, P, sc) -> (3,3), where P has shape (3,3).
+    eps : float
+        Numerical safeguard used for divisions/inversions.
+    """
+
+    preset: MetricPreset = "identity"
+    diagonal_weights: dict[int, np.ndarray] | None = None
+    primal_measures: dict[int, np.ndarray] | None = None
+    dual_measures: dict[int, np.ndarray] | None = None
+    tensors: np.ndarray | None = None
+    fn: MetricCallable | None = None
+    eps: float = 1e-12
+
+
+class HodgeStarBackend(Protocol):
+    """Protocol for Hodge star backends."""
+
+    def star(self, sc: Any, k: int, *, inverse: bool = False) -> csr_matrix:
+        """Return *_k (or its inverse).
+
+        Parameters
+        ----------
+        sc : Any
+            Complex object.
+        k : int
+            Cochain degree.
+        inverse : bool
+            If True, return (*_k)^{-1}.
+
+        Returns
+        -------
+        csr_matrix
+            Sparse star matrix.
+        """
+
+
+def _as_1d_float(x: np.ndarray, *, name: str) -> np.ndarray:
```

Comment:
It feels weird to require a `name` parameter just to be used in an error message. Whats the point? Context should be sufficiently given in the stacktrace?